src/blog/2024/11/esp32-with-node-red.md

```diff
@@ -0,0 +1,258 @@
+{% endrenderFlow %}
+
+Now, when you click either the "ON" or "OFF" button on the dashboard, it will send either 1 or 2 as the payload. The ESP32 will use this payload in its code to turn the LED on or off. To view the dashboard, switch to the Dashboard 2.0 tab on the right side and click the Open Dashboard button. The dashboard will look similar to the image below.
+
+![FlowFuse Dashboard Build to control the ESP32 LED](./images/dashboard2.png){data-zoomable}
+_FlowFuse Dashboard Build to control the ESP32 LED_
+
+### Step 4: Programming ESP32 to receive commands from MQTT and Control LED
+
+Now, let's move on to the final step. Before proceeding, make sure your ESP32 is **connected to your laptop or computer via USB***. The USB connection is essential for uploading the code (sketch) to the ESP32, which will enable it to connect to the internet and communicate with the MQTT broker.
+The ESP32 will subscribe to the MQTT topic we configured earlier (e.g., /LEDControl). Based on the received payload (1 or 2), it will control the LED accordingly — turning it on or off.
+
+**Setting up Arduino IDE:**
+
+1. Open the Arduino IDE on your computer.
+2. Ensure you have selected the correct board and port in the Tools menu.
+3. Install the necessary library:
+4. Go to Tools > "Manage Libraries".
+5. Search for and install the "EspMQTTClient" library by Patrick Lapointe.
+6. The library installation will prompt you to install its dependencies—ensure that you tick that option and proceed to install.
+
+**Code for ESP32:**
+
+1. Copy the following code into the Arduino IDE:
+
+```cpp
+#if defined(ESP32)
+#include <WiFi.h>
+#elif defined(ESP8266)
+#include <ESP8266WiFi.h>
+#endif
+
+#include <PubSubClient.h>
+
+#define LedPin 2 // ESP32 built-in LED pin
+
+// WiFi and MQTT settings
+const char* ssid = ""; // Change this to your WiFi SSID
+const char* password = ""; // Change this to your WiFi password
+const char* mqtt_server = "broker.flowfuse.cloud"; // FlowFuse MQTT broker server
+
+// MQTT client credentials
+const char* mqtt_client_id = ""; // Replace with your MQTT client ID
+const char* mqtt_username = ""; // Replace with your MQTT username
+const char* mqtt_password = ""; // Replace with your MQTT password
+
+WiFiClient espClient;
+PubSubClient client(espClient);
+
+// Function to connect to WiFi
+void setup_wifi() {
+    delay(10);
```

Comment:
Curious as to why the delay?

Reply:
By adding a delay, we allow a brief moment for the serial communication to properly initialize before printing the first message. However, this delay is not strictly necessary we can remove it.